src/sources/splunk_hec.rs

```diff
@@ -0,0 +1,954 @@
+                } else {
+                    return Err(event_error("Invalid data format", 6, self.events));
+                }
+            }
+            Some(None) => {
+                return Err(event_error("Invalid data format", 6, self.events));
+            }
+        }
+
+        // Add time field
+        if let Some(time) = self.time.as_ref() {
+            log.insert_explicit(event::TIMESTAMP.clone(), time.clone());
+        }
+
+        // Extract default extracted fields
+        for de in self.extractors.iter_mut() {
+            de.extract(log, &mut json);
+        }
+
+        self.events += 1;
+
+        Ok(Async::Ready(Some(event)))
+    }
+}
+
+/// Maintains last known extracted value of field and uses it in the absence of field.
+struct DefaultExtractor {
+    field: &'static Atom,
+    value: Option<ValueKind>,
+}
+
+impl DefaultExtractor {
+    fn new(field: &'static Atom) -> Self {
+        DefaultExtractor { field, value: None }
+    }
+
+    fn new_with(field: &'static Atom, value: impl Into<Option<ValueKind>>) -> Self {
+        DefaultExtractor {
+            field,
+            value: value.into(),
+        }
+    }
+
+    fn extract(&mut self, log: &mut LogEvent, value: &mut Value) {
+        // Process json_field
+        if let Some(Value::String(new_value)) = value.get_mut(self.field.as_ref()).map(Value::take)
+        {
+            self.value = Some(new_value.into());
+        }
+
+        // Add data field
+        if let Some(index) = self.value.as_ref() {
+            log.insert_explicit(self.field.clone(), index.clone());
+        }
+    }
+}
+
+/// Creates event from raw request
+fn raw_event(
+    bytes: Bytes,
+    gzip: bool,
+    channel: HeaderValue,
+    host: Option<HeaderValue>,
+) -> Result<Event, Response<Body>> {
+    // Process gzip
+    let bytes = if gzip {
+        let mut data = Vec::new();
+        match GzDecoder::new(bytes.into_buf()).read_to_end(&mut data) {
+            Ok(0) => {
+                return Err(response(
+                    StatusCode::BAD_REQUEST,
+                    splunk_response::NO_DATA.clone(),
+                ))
+            }
+            Ok(_) => data.into(),
+            Err(error) => {
+                error!(message = "Malformed request body",%error);
+                return Err(event_error("Invalid data format", 6, 0));
+            }
+        }
+    } else {
+        bytes
+    };
+
+    // Construct event
+    let mut event = Event::new_empty_log();
+    let log = event.as_mut_log();
+
+    // Add message
+    log.insert_explicit(event::MESSAGE.clone(), bytes.into());
+
+    // Add channel
+    log.insert_explicit(CHANNEL.clone(), channel.as_bytes().into());
+
+    // Add host
+    if let Some(host) = host {
+        log.insert_explicit(event::HOST.clone(), host.as_bytes().into());
+    }
+
+    Ok(event)
+}
+
+/// Recursevly inserts json values to event under given name
+pub fn insert(event: &mut LogEvent, name: String, value: Value) {
```

Comment:
This seems generally useful, might be worth placing this within the event module as a method. It'd also make it easier to find when https://github.com/timberio/vector/issues/704 gets worked on.

Reply:
I agree. 

Also this is actually a modified version of basically the same function in `json_parser`, so I will also adjust `json_parser` to use that.